File: app/src/main/java/com/privat/contacts/domain/model/UserDomain.java
package com.privat.contacts.domain.model;

import org.jetbrains.annotations.NotNull;

import java.util.LinkedList;
import java.util.List;

public interface UserDomain {
    int id();

    @NotNull
    String photoUrl();

    @NotNull
    <T> T map(Mapper<T> mapper);

    static <T> List<T> mapList(List<UserDomain> userDomainList, Mapper<T> mapper) {
        LinkedList<T> newList = new LinkedList();
        for (UserDomain item :
                userDomainList) {
            newList.add(item.map(mapper));
        }
        return newList;
    }

    interface Mapper<T> {
        @NotNull
        T map(int id,
              String uid,
              String password,
              String firstName,
              String lastName,
              String username,
              String email,
              String avatar,
              String gender,
              String phoneNumber,
              String socialInsuranceNumber,
              String dateOfBirth,
              String cardNumber,
              boolean favorite,
              AddressDomain address,
              EmploymentDomain employment,
              SubscriptionDomain subscription);
    }

    boolean sameId(int userId);
}

File: app/src/main/java/com/privat/contacts/domain/model/UserDomain.kt
package com.privat.contacts.domain.model

import java.util.LinkedList

interface UserDomain {
    fun id(): Int
    fun photoUrl(): String
    fun <T> map(mapper: Mapper<T>): T
    interface Mapper<T> {
        fun map(id: Int,
                uid: String,
                password: String,
                firstName: String,
                lastName: String,
                username: String,
                email: String,
                avatar: String,
                gender: String,
                phoneNumber: String,
                socialInsuranceNumber: String,
                dateOfBirth: String,
                cardNumber: String,
                favorite: Boolean,
                address: AddressDomain,
                employment: EmploymentDomain,
                subscription: SubscriptionDomain): T
    }

    fun sameId(userId: Int): Boolean

    companion object {
        fun <T> mapList(userDomainList: List<UserDomain>, mapper: Mapper<T>): List<T> {
            val newList: LinkedList<T> = LinkedList()
            for (item in userDomainList) {
                newList.add(item!!.map(mapper))
            }
            return newList
        }
    }
}
